### PR TITLE
fix: Multitenant session

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

- Link to issue1 here
- Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/main/java/io/supertokens/session/Session.java [hunk 1]
```diff
@@ -229,17 +236,18 @@ public static SessionInformationHolder getSession(AppIdentifier appIdentifier, M
         AccessTokenInfo accessToken = AccessToken.getInfoFromAccessToken(appIdentifier, main,
                 token,
                 doAntiCsrfCheck && enableAntiCsrf);
-        TenantIdentifier tenantIdentifier = accessToken.tenantIdentifier;
+        TenantIdentifierWithStorage tenantIdentifierWithStorage = accessToken.tenantIdentifier.withStorage(
+                StorageLayer.getStorage(accessToken.tenantIdentifier, main));
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/Session.java:241]: check if the input app identifier has the same appid as the one from the access token in the getInfoFromAccessToken function

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java [hunk 2]
```diff
@@ -155,12 +156,18 @@ public static AccessTokenInfo getInfoFromAccessToken(AppIdentifier appIdentifier
...
-    public static AccessTokenInfo getInfoFromAccessTokenWithoutVerifying(@Nonnull String token) {
-        return getInfoFromAccessTokenWithoutVerifying(new AppIdentifier(null, null), token);
+    public static AccessTokenInfo getInfoFromAccessTokenWithoutVerifying(Main main, @Nonnull String token) {
+        try {
+            return getInfoFromAccessTokenWithoutVerifying(
+                    new AppIdentifier(null, null), main, token);
+        } catch (TenantOrAppNotFoundException e) {
+            throw new IllegalStateException(e);
+        }
     }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/accessToken/AccessToken.java:167]: In the AccessTokenPayload, we store the tenantId + another field called `appIdHash: hash(connectionUrl + appid)`

Whenever we decode / fetch info from the accessToken, we check if the hash of the input appIdentifier is the same as the hash in the access token payload. If not, we throw a badpermissionexception

File: src/main/java/io/supertokens/session/Session.java [hunk 3]
```diff
@@ -171,17 +177,18 @@ public static SessionInformationHolder regenerateToken(AppIdentifier appIdentifi
...
         // changed for it...
-        AccessTokenInfo accessToken = AccessToken.getInfoFromAccessTokenWithoutVerifying(appIdentifier, token);
-        TenantIdentifier tenantIdentifier = accessToken.tenantIdentifier;
-        if (!tenantIdentifier.toAppIdentifier().equals(appIdentifier)) {
+        AccessTokenInfo accessToken = AccessToken.getInfoFromAccessTokenWithoutVerifying(appIdentifier, main, token);
+        TenantIdentifierWithStorage tenantIdentifierWithStorage = accessToken.tenantIdentifier.withStorage(
+                StorageLayer.getStorage(accessToken.tenantIdentifier, main));
+        if (!tenantIdentifierWithStorage.toAppIdentifier().equals(appIdentifier)) {
             throw new UnauthorisedException("Access token is from an incorrect app");
         }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/Session.java:185]: this check goes inside getInfoFromAccessTokenWithoutVerifying

File: src/main/java/io/supertokens/session/Session.java [hunk 4]
```diff
@@ -423,11 +430,14 @@ public static SessionInformationHolder refreshSession(AppIdentifier appIdentifie
             }
         }
 
-        return refreshSessionHelper(refreshTokenInfo.tenantIdentifier, main, refreshToken, refreshTokenInfo,
+        return refreshSessionHelper(
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/session/Session.java:433]: the function for getInfoFromRefreshToken should also store the appIdHash and do a check similar to in the access token

File: src/main/java/io/supertokens/webserver/api/session/HandshakeAPI.java [hunk 5]
```diff
@@ -49,35 +50,34 @@ public String getPath() {
...
                             .getKeyExpiryTime());
 
             if (!super.getVersionFromRequest(req).equals("2.7") && !super.getVersionFromRequest(req).equals("2.8")) {
-                List<KeyInfo> keys = AccessTokenSigningKey.getInstance(this.getTenantIdentifierWithStorageFromRequest(req).toAppIdentifier(),
-                                main)
+                List<KeyInfo> keys = AccessTokenSigningKey.getInstance(tenantIdentifier.toAppIdentifier(), main)
                         .getAllKeys();
                 JsonArray jwtSigningPublicKeyListJSON = Utils.keyListToJson(keys);
                 result.add("jwtSigningPublicKeyList", jwtSigningPublicKeyListJSON);
             }
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/session/HandshakeAPI.java:74]: are you sure all the props here are not per app?

File: src/main/java/io/supertokens/session/accessToken/AccessToken.java (hunk 2 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/accessToken/AccessToken.java:172]: Done, throwing unauthorized exception (like it was doing before)

File: src/main/java/io/supertokens/session/Session.java (hunk 3 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/Session.java]: Done

File: src/main/java/io/supertokens/session/Session.java (hunk 4 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/session/Session.java:431]: Done

File: src/main/java/io/supertokens/webserver/api/session/HandshakeAPI.java (hunk 5 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/session/HandshakeAPI.java:74]: properties seems to be fine

**Pull request merged**